src/Microsoft.DotNet.CodeAnalysis/Analyzers/PinvokeAnalyzer.cs

```diff
@@ -0,0 +1,157 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.IO;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Text;
+using System.Diagnostics;
+
+namespace Microsoft.DotNet.CodeAnalysis.Analyzers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class PinvokeAnalyzer : DiagnosticAnalyzer
+    {
+        private const string Title = "Invalid Pinvoke call";
+        private const string MessageFormat = @"{0} is not supported on one\more targeted platforms.{1}";
+        private const string Description = "The Pinvoke call should be removed.";
+        private const string AnalyzerName = "PinvokeAnalyzer";
+
+        private static DiagnosticDescriptor InvalidPinvokeCall = new DiagnosticDescriptor(DiagnosticIds.BCL0100.ToString(), Title, MessageFormat, AnalyzerName, DiagnosticSeverity.Error, isEnabledByDefault: true, description: Description);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get { return ImmutableArray.Create(InvalidPinvokeCall); } }
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private void OnCompilationStart(CompilationStartAnalysisContext obj)
+        {
+            AllowedPinvokeFile = obj.Options.AdditionalFiles.FirstOrDefault(f => Path.GetFileName(f.Path).Contains("PinvokeAnalyzer_"));
+            ExceptionFile = obj.Options.AdditionalFiles.FirstOrDefault(f => Path.GetFileName(f.Path).Contains("PinvokeAnalyzerExceptionList.analyzerdata"));
+            obj.RegisterSymbolAction(AnalyzeMethod, SymbolKind.Method);
+        }
+
+        private Dictionary<string, string> _allowedPinvokes;
+        private Dictionary<string, string> _exceptionPinvokes;
+
+        private Dictionary<string, string> AllowedPinvokes
+        {
+            get
+            {
+                if (_allowedPinvokes == null)
+                {
+                    _allowedPinvokes = ParseAdditionalFile(AllowedPinvokeFile);
+                }
+
+                return _allowedPinvokes;
+            }
+        }
+
+        private Dictionary<string, string> ExceptionPinvokes
+        {
+            get
+            {
+                if (_exceptionPinvokes == null)
+                {
+                    _exceptionPinvokes = ParseAdditionalFile(ExceptionFile);
+                }
+
+                return _exceptionPinvokes;
+            }
+        }
+
+        private AdditionalText ExceptionFile { get; set; }
+        private AdditionalText AllowedPinvokeFile { get; set; }
+
+        private HashSet<string> _isNotSupportedOnWin7 = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+        private HashSet<string> IsNotSupportedOnWin7 { get { return _isNotSupportedOnWin7; } }
```

Comment:
In case we need extra logic I think we should add it then. This is going to be called in a hot path during compilation and method calls adds to the overhead. I would prefer we keep these as fields until we need more logic around them.